### ISSUE TITLE
[HOLD #48400][$250] Expense - RBR doesn't disappear Instantly after paying held expense

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#48397]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.27-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** biruknew45+518@gmail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. As User A, go to the chat with User B.
3. As User A, submit an expense.
4. As User A, right-click on the expense and click "Hold."
5. The RBR appears in the expense preview for both User A and User B. If not, go to the expense details and then return to the chat.
6. As User B, click "Pay Elsewhere" and confirm the payment.
7. Click on the expense preview and then return to the chat (repeat this for both User A and User B).

## Expected Result:
After paying the held expense in step 6, the RBR in the preview should disappear instantly.

## Actual Result:
After paying the held expense in step 6, the RBR in the preview persists. The red dot only disappears after clicking on the preview and returning to the chat.

## Workaround:
Unknown

## Platforms:

- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

![Bug6589914_1725218970132!1 (1)](https://github.com/user-attachments/assets/c5c0a277-93f4-4bd7-9e94-2d2b8671d573)

https://github.com/user-attachments/assets/00f5ba3a-6044-4ed1-af7b-bfb4f30bbb2d

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021831729851068678041</li>
        <li>Upwork Job ID: 1831729851068678041</li>
        <li>Last Price Increase: 2024-09-19</li>
<li>Automatic offers: </li>
<ul>
<li>ikevin127 | Reviewer | 104042668</li>
<li>nkdengineer | Contributor | 104042669</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @twisterdotcom</details>

**Comment by IuliiaHerets** [comment]
We think that this bug might be related to #vip-bills

**Comment by IuliiaHerets** [comment]
@twisterdotcom  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by twisterdotcom** [comment]
Yes, I recreated this.

https://github.com/user-attachments/assets/346c33cd-e9a0-44b8-87a4-765656829786

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After paying the held expense in step 6, the RBR in the preview persists. The red dot only disappears after clicking on the preview and returning to the chat.

### What is the root cause of that problem?
When we pay money request, we unhold all transactions. But we don't clear the hold violation in optimistic data and after the API is complete, BE also doesn't clear the hold violation of the transaction then RBR still displays.


https://github.com/Expensify/App/blob/65c0f4103f178bf18f81a8b4e71ace5203c912cf/src/libs/actions/IOU.ts#L6725-L6734

This bug also happens when we approve money request but it only happens in offline, after the API is complete, BE returns violation of transaction that cleared the hold violation

https://github.com/Expensify/App/blob/65c0f4103f178bf18f81a8b4e71ace5203c912cf/src/libs/actions/IOU.ts#L7018-L7023
### What changes do you think we should make in order to solve the problem?

We should clear the hold violation when we pay/approve all requests. To prevent duplicate code, we can create a util to get the optimistic data for this case

```
function buildOnyxDataForUnHoldTransaction(expenseReport: OnyxEntry<OnyxTypes.Report>) {
    const optimisticData: OnyxUpdate[] = [];
    const failureData: OnyxUpdate[] = [];

    const heldTransactions = ReportUtils.getAllHeldTransactions(expenseReport?.reportID);
    heldTransactions.forEach((heldTransaction) => {
        optimisticData.push({
            onyxMethod: Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.TRANSACTION}${heldTransaction.transactionID}`,
            value: {
                comment: {
                    hold: '',
                },
            },
        });
        failureData.push({
            onyxMethod: Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.TRANSACTION}${heldTransaction.transactionID}`,
            value: {
                comment: {
                    hold: heldTransaction.comment?.hold,
                },
            },
        });
        const transactionViolations = allTransactionViolations[`${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${heldTransaction.transactionID}`] ?? []
        optimisticData.push({
            onyxMethod: Onyx.METHOD.SET,
            key: `${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${heldTransaction.transactionID}`,
            value: transactionViolations.filter((violation) => violation.name !== CONST.VIOLATIONS.HOLD)
        });
        failureData.push({
            onyxMethod: Onyx.METHOD.SET,
            key: `${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${heldTransaction.transactionID}`,
            value: transactionViolations,
        });
    });

    return {optimisticData, failureData};
}
```
Then we can use this in both places here

```
if (full) {
    const unholdTransactionOnyxData = buildOnyxDataForUnHoldTransaction(iouReport);
    optimisticData.push(...unholdTransactionOnyxData.optimisticData);
    failureData.push(...unholdTransactionOnyxData.failureData);
}
```

https://github.com/Expensify/App/blob/65c0f4103f178bf18f81a8b4e71ace5203c912cf/src/libs/actions/IOU.ts#L6725-L6734

```
if (full && hasHeldExpenses) {
    const unholdTransactionOnyxData = buildOnyxDataForUnHoldTransaction(expenseReport);
    optimisticData.push(...unholdTransactionOnyxData.optimisticData);
    failureData.push(...unholdTransactionOnyxData.failureData);
}
```
https://github.com/Expensify/App/blob/65c0f4103f178bf18f81a8b4e71ace5203c912cf/src/libs/actions/IOU.ts#L7018-L7023

OPTIONAL: BE should also clear the hold violation when returning the data in `PayMoneyRequest` API
### What alternative solutions did you explore? (Optional)

**Comment by twisterdotcom** [comment]
@abdulrahuman5196 how are we doing on the reviews here?

**Comment by abdulrahuman5196** [comment]
@twisterdotcom I am multiple items on my plate. Kindly assign it to a different C+. Unassigning myself.

**Comment by twisterdotcom** [comment]
@ikevin127 could you review the proposal here from @nkdengineer and @Nodebrute?

**Comment by ikevin127** [comment]
@twisterdotcom Sure, will review today.

**Comment by ikevin127** [comment]
@nkdengineer's [proposal](https://github.com/Expensify/App/issues/48397#issuecomment-2333212238) looks good to me.

**TLDR:** Fixing this issue 100% **does** require BE changes as well.

> [!important]
> The RCA is correct and the proposed solution solves our issue (**partially**) in the sense that the RBR does disappear on **User B**'s side when the held expense is paid because the onyx data is only changed locally (in the browser), hence why it's not removed from **User A** because that would require BE to send a pusher event with the required data to remove the RBR after **User B** pays the held expense.

🎀👀🎀 C+ reviewed

**Comment by ikevin127** [comment]
@Nodebrute Thanks for the proposal. Unfortunately the proposed solution does not work as `transaction` is undefined which besides not fixing our issue, it blocks the payment when trying to confirm Pay elsewhere, because of the undefined variable mentioned above.

<img width="1315" alt="error" src="https://github.com/user-attachments/assets/dcd62a16-df7e-4215-b127-a587a3d9ebf6">

The solution is not complete enough to facilitate reviewing / testing.

**Comment by thienlnam** [comment]
cc @cead22 on the held violation here. It looks like on an paid request, the violations on the transaction do not clear. Is this something we should take care of in Web-E or wait until the new violations roll out?

**Comment by cead22** [comment]
@thienlnam this might be a duplicate of https://github.com/Expensify/App/issues/48400. I think we should fix it now, and if it's a dupe we can handle it in that other issue

**Comment by thienlnam** [comment]
Ah yeah, actually this does look like a dupe and the other issue already has an assignee

**Comment by nkdengineer** [comment]
@cead22 I think we should clear the hold violation in optimistic data if we paid full instead of adding `!isSettled` condition.

**Comment by ikevin127** [comment]
There's a 3rd dupe of this issue that might be fixed in similar fashion, only difference being it's a split instead of hold:
- https://github.com/Expensify/App/issues/48423

**Comment by thienlnam** [comment]
@ikevin127 That looks like it would also be solved by https://github.com/Expensify/App/issues/48400

The steps don't look reproducible because it's probably because the expense needs to have a violation

**Comment by twisterdotcom** [comment]
#48400 is closed now so I'll try to retest this.

**Comment by ikevin127** [comment]
Retesting.

**Comment by twisterdotcom** [comment]
Yeah, fixed.

https://github.com/user-attachments/assets/74d125c0-3147-4d91-9c3f-5eed0104daff
